iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnTensors.cpp

```diff
@@ -88,30 +65,49 @@ struct TorchIndexSelectOpConversion
     mhlo::TorchIndexSelectOp::Adaptor adaptor(args);
     int axis = op.dim();
     int batch = op.batch_dims();
-    auto indexShapeType = adaptor.index().getType().dyn_cast<ShapedType>();
+    auto indexShapeType = adaptor.index().getType().cast<ShapedType>();
     int nIndices = indexShapeType.getRank();
-    auto inputShapeType = adaptor.input().getType().dyn_cast<ShapedType>();
+    auto inputShapeType = adaptor.input().getType().cast<ShapedType>();
     if (axis < 0) axis += inputShapeType.getRank();
     if (batch < 0) batch += nIndices;
 
     Location loc = op.getLoc();
-    Value output = op.getResult();
-    int rank = output.getType().cast<ShapedType>().getRank();
+    ShapedType resultType = op.getResult().getType().cast<ShapedType>();
+    int rank = resultType.getRank();
+
     SmallVector<AffineMap, 2> indexingMaps;
     SmallVector<AffineExpr, 4> exprs;
-    for (int i = 0; i < batch; ++i)
+    for (int i = 0; i < batch; ++i) {
       exprs.push_back(rewriter.getAffineDimExpr(i));
-    for (int i = 0, e = nIndices - batch; i < e; ++i)
+    }
+    for (int i = 0, e = nIndices - batch; i < e; ++i) {
       exprs.push_back(rewriter.getAffineDimExpr(axis + i));
+    }
     indexingMaps.emplace_back(
         AffineMap::get(rank, /*symbolCount=*/0, exprs, rewriter.getContext()));
     indexingMaps.emplace_back(rewriter.getMultiDimIdentityMap(rank));
+
     SmallVector<StringRef, 3> loopTypes(rank, getParallelIteratorTypeName());
-    ShapedType outputType = op.getResult().getType().cast<ShapedType>();
-    Value initOp = getOutputTensor(rewriter, loc, op.getResult());
-    if (!initOp) return failure();
+
+    // The output shape is
+    //   `params[:axis] + indices[batch_dims:] + params[axis + 1:]`
+    SmallVector<Value, 4> dynSizes;
+    for (int i = 0; i < rank; ++i) {
+      if (!resultType.isDynamicDim(i)) continue;
+      if (i < axis) {
+        dynSizes.push_back(rewriter.create<DimOp>(loc, adaptor.input(), i));
+      } else if (i < (axis + nIndices - batch)) {
+        int idx = i - axis + batch;
+        dynSizes.push_back(rewriter.create<DimOp>(loc, adaptor.index(), idx));
+      } else {
+        int idx = i - (axis + nIndices - batch) + axis + 1;
+        dynSizes.push_back(rewriter.create<DimOp>(loc, adaptor.input(), idx));
+      }
+    }
+    Value initOp = rewriter.create<linalg::InitTensorOp>(
```

Comment:
Not required, but `InitTensorOp` needs a builder method that can take `ArrayRef<OpFoldResult>` for the shape. Bonus points for doing that upstream and using that here.

Reply:
InitTensorOp has a builder method that can take `ArrayRef<OpFoldResult>`. But this doesn't work to me. I feel createOrFold returns a Value, so it will not be treated as an Attribute. So it is actually storing a ConstantOp, not Attribute.

I feel the better way is to fix `dispatchIndexOpFoldResult` in LinalgOps.cpp to handle the case of ConstantOp.

Reply:
I tried, but the fix would be SUPER painful. It breaks a lot of things, including bunch of integration tests.

Reply:
Sure NP. Ill look into it later. I didnt think it would hit these issues. Thanks for trying